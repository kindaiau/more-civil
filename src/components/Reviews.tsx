import { useEffect, useState } from 'react';

type Review = {
  author_name: string
  profile_photo_url?: string
  relative_time_description?: string
  text: string
  rating: number
  time?: number
}

export default function ReviewsSection() {
  const [data, setData] = useState<{
    rating?: number
    user_ratings_total?: number
    reviews: Review[]
    placePhotoRef?: string | null
    error?: string
  }>({ reviews: [] })

  useEffect(() => {
    fetch('/api/google-reviews')
      .then((r) => r.json())
      .then(setData)
      .catch(() => setData({ reviews: [], error: 'fetch-failed' }))
  }, [])

  useEffect(() => {
    if (!data) return
    const existing = document.getElementById('ld-localbusiness')
    if (existing) existing.remove()
    const rating = typeof data.rating === 'number' ? data.rating : 5
    const count = typeof data.user_ratings_total === 'number' ? data.user_ratings_total : 3
    const script = document.createElement('script')
    script.id = 'ld-localbusiness'
    script.type = 'application/ld+json'
    script.textContent = JSON.stringify({
      '@context': 'https://schema.org',
      '@type': 'LocalBusiness',
      name: 'More Civil',
      address: { '@type': 'PostalAddress', addressRegion: 'SA', addressCountry: 'AU' },
      url: 'https://more-civil.lovable.app/',
      aggregateRating: { '@type': 'AggregateRating', ratingValue: rating, reviewCount: count },
      ...(data.reviews.length ? {
        review: data.reviews.map(r => ({
          '@type': 'Review',
          author: r.author_name,
          reviewBody: r.text,
          reviewRating: { '@type': 'Rating', ratingValue: r.rating }
        }))
      } : {})
    })
    document.head.appendChild(script)
    return () => {
      try {
        const scriptToRemove = document.getElementById('ld-localbusiness')
        if (scriptToRemove) {
          scriptToRemove.remove()
        }
      } catch (error) {
        console.warn('Failed to remove JSON-LD script:', error)
      }
    }
  }, [data])

  const placeImg = data.placePhotoRef
    ? `/api/google-photo?ref=${encodeURIComponent(data.placePhotoRef)}&maxwidth=800`
    : undefined

  return (
    <section className="py-16 bg-white">
      <div className="mx-auto max-w-screen-xl px-4 sm:px-6 lg:px-8">
        <h2 className="text-2xl sm:text-3xl font-semibold tracking-tight">What our clients say</h2>
        {typeof data.rating === 'number' && typeof data.user_ratings_total === 'number' && (
          <p className="mt-2 text-sm text-neutral-600">
            Google rating {data.rating.toFixed(1)} / 5 from {data.user_ratings_total} reviews
          </p>
        )}
        <div className="mt-8 grid grid-cols-1 md:grid-cols-3 gap-6">
          {data.reviews?.length ? (
            data.reviews.map((r, i) => (
              <article key={i} className="rounded-2xl border border-neutral-200 p-6 shadow-sm">
                <div className="flex items-center gap-3">
                  {r.profile_photo_url ? (
                    <img src={r.profile_photo_url} alt={`${r.author_name} profile`} className="w-10 h-10 rounded-full object-cover" />
                  ) : placeImg ? (
                    <img src={placeImg} alt="More Civil project" className="w-10 h-10 rounded-full object-cover" />
                  ) : (
                    <div className="w-10 h-10 rounded-full bg-neutral-200" />
                  )}
                  <div>
                    <div className="text-sm font-medium">{r.author_name}</div>
                    <div className="text-xs text-neutral-600">{r.relative_time_description || ''}</div>
                  </div>
                </div>
                <div className="mt-3 text-amber-500" aria-label={`${r.rating} out of 5 stars`}>
                  {'★★★★★'.slice(0, r.rating)}
                </div>
                <p className="mt-3 text-neutral-800">{r.text}</p>
              </article>
            ))
          ) : (
            <>
              {[
                'Reliable water delivery when we needed it most.',
                'Professional team. Fast response. Great on-site communication.',
                'Five stars. Job done right and on time.',
              ].map((t, i) => (
                <article key={i} className="rounded-2xl border border-neutral-200 p-6 shadow-sm">
                  <div className="text-amber-500" aria-label="5 out of 5 stars">★★★★★</div>
                  <p className="mt-3 text-neutral-800">{t}</p>
                  <div className="mt-3 text-sm text-neutral-600">Google reviewer</div>
                </article>
              ))}
            </>
          )}
        </div>
      </div>
    </section>
  )
}
