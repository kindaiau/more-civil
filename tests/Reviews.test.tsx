import { describe, it, expect, vi } from 'vitest';
import { render } from '@testing-library/react';
import Reviews from '../src/components/Reviews';

describe('Reviews component', () => {
  it('renders fetched reviews', async () => {
    vi.stubGlobal('fetch', () =>
      Promise.resolve({
        json: () =>
          Promise.resolve({
            reviews: [
              {
                author_name: 'Jane',
                profile_photo_url: 'avatar.jpg',
                rating: 5,
                text: 'Great work!',
                time: 123,
                original_url_to_review: 'https://example.com',
                photo: { thumb: 'thumb.jpg', full: 'full.jpg' },
              },
            ],
          }),
      })
    );
    const { findByText, container } = render(<Reviews />);
    await findByText('Great work!');
    expect(container).toMatchSnapshot();
  });
});
