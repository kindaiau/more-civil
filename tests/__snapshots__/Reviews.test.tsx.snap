// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Reviews component > renders fetched reviews 1`] = `
<div>
  <section
    aria-labelledby="reviews-title"
    class="py-12 sm:py-16 lg:py-20"
    id="reviews"
  >
    <div
      class="mx-auto max-w-screen-xl px-4 sm:px-6 lg:px-8"
    >
      <h2
        class="text-3xl font-extrabold text-center mb-8"
        id="reviews-title"
        style="font-family: Montserrat;"
      >
        Customer Reviews
      </h2>
      <div
        class="grid gap-8 md:grid-cols-3"
      >
        <article
          class="bg-white border border-[#E5E7EB] rounded-lg shadow-sm p-5"
        >
          <div
            class="flex items-center mb-4"
          >
            <img
              alt="Jane's profile photo"
              class="w-14 h-14 rounded-full object-cover"
              loading="lazy"
              src="avatar.jpg"
            />
            <div
              class="ml-4"
            >
              <p
                class="font-semibold"
              >
                Jane
              </p>
              <div
                aria-label="5 star rating"
                class="flex"
              >
                <svg
                  aria-hidden="true"
                  class="lucide lucide-star w-4 h-4 text-[#00B4D8] fill-[#00B4D8]"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="lucide lucide-star w-4 h-4 text-[#00B4D8] fill-[#00B4D8]"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="lucide lucide-star w-4 h-4 text-[#00B4D8] fill-[#00B4D8]"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="lucide lucide-star w-4 h-4 text-[#00B4D8] fill-[#00B4D8]"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="lucide lucide-star w-4 h-4 text-[#00B4D8] fill-[#00B4D8]"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <p
            class="text-sm text-slate-700"
            style="overflow: hidden;"
          >
            Great work!
          </p>
          <a
            class="mt-4 block w-[120px] h-[120px]"
            href="full.jpg"
            rel="noopener"
            target="_blank"
          >
            <img
              alt="Customer submitted photo"
              class="w-[120px] h-[120px] object-cover rounded"
              loading="lazy"
              src="thumb.jpg"
            />
          </a>
          <a
            class="mt-4 inline-block text-sm text-[#00B4D8]"
            href="https://example.com"
            rel="noopener"
            target="_blank"
          >
            Read on Google
          </a>
        </article>
      </div>
      <div
        class="text-center mt-8"
      >
        <a
          class="text-sm text-[#00B4D8]"
          href="https://share.google/5FTrL7ap7qMBWKyVz"
          rel="noopener"
          target="_blank"
        >
          See all reviews on Google â†’
        </a>
      </div>
    </div>
  </section>
</div>
`;
